filebeat/harvester/log.go

```diff
@@ -294,11 +295,34 @@ func (h *Harvester) newLogFileReader() (reader.Reader, error) {
 	// TODO: NewLineReader uses additional buffering to deal with encoding and testing
 	//       for new lines in input stream. Simple 8-bit based encodings, or plain
 	//       don't require 'complicated' logic.
-	fileReader, err := NewLogFile(h.file, h.config, h.done)
+	fileReader, err := NewLogFile(h.file, h.config)
 	if err != nil {
 		return nil, err
 	}
 
+	// Makes sure file reader is closed when done channel is closed
+	go func() {
+		select {
+		case <-h.done:
+			fileReader.Close()
+			return
+		}
+	}()
+
+	// Closes reader after timeout
+	if h.config.CloseTimeout > 0 {
+		go func() {
+			select {
+			case <-h.done:
+				return
+			case <-time.After(h.config.CloseTimeout):
+				logp.Info("Closing harvester because close_timeout was reached: %s", h.state.Source)
+				fileReader.Close()
+				return
+			}
+		}()
+	}
+
```

Comment:
alternative implementation:

```
var closeTimeout chan time.Time
if h.config.CloseTimeout > 0 {
    closeTimeout = time.After(h.config.CloseTimeout):
}
go func () {
    select {
        case <- h.done:
        case <- closeTimeout:
            logp.Info("Closing harvester because close_timeout was reached: %s", h.state.Source)
    }
    fileReader.Close()
}
```